core/templates/pages/exploration-editor-page/editor-tab/services/responses.service.ts

```diff
@@ -323,10 +323,60 @@ angular.module('oppia').factory('ResponsesService', [
       // Updates answer choices when the interaction requires it -- for
       // example, the rules for multiple choice need to refer to the multiple
       // choice interaction's customization arguments.
-      updateAnswerChoices: function(newAnswerChoices, callback) {
+      updateAnswerChoices: function(
+          newAnswerChoices, oldToNewAnswerChoicesMapping, callback) {
         var oldAnswerChoices = angular.copy(_answerChoices);
         _answerChoices = newAnswerChoices;
 
+        // If the interaction is MultipleChoiceInput, update the answer groups
+        // to refer to the new answer options.
+        if (StateInteractionIdService.savedMemento === 'MultipleChoiceInput') {
+          // Save and update the answer groups for MultipleChoiceInput
+          // based on the information about the modification of answer choices
+          // from the schema-based-list editor.
+          // The information regarding the modifications of answer choices is
+          // present in oldToNewAnswerChoicesMapping.
+
+          if (
+            oldToNewAnswerChoicesMapping !== undefined &&
+               oldToNewAnswerChoicesMapping !== null &&
+               oldToNewAnswerChoicesMapping.hasOwnProperty('deletedIndexes')) {
+            var deletedIndexes = oldToNewAnswerChoicesMapping.deletedIndexes;
+            var answerGroupsBeforeUpdating = angular.copy(_answerGroups);
```

Comment:
I know you are doing this because you are updating _answerGroups. But given that you are never changing the order of elements in answerGroupsBeforeUpdating or in answerGroupsBeforeUpdating[i].rules, could you explain exactly what goes wrong if you don't do a copy?

Reply:
Assume we didn't make a copy of the `_answerGroups`.
The `deletedIndexes` are indexes of the choice which are missing from the old Answer choices.
So for example, Suppose there are 5 choices and choices 1 and 3 correspond to an answer(The index of the rules are then 0 and 2). let the `deletedIndexes` also be 0 and 2. 
Now for the first element of `deletedIndexes` i.e 0, in the loop `for (var j = 0; j < rules.length; j++)`
both if and else if conditions are satisfied. The first rule index is made invalid(MAX_INTEGER) and the second
rule is reduced by one. Now after this the new `_answerGroups` rules are 0 and 1.
But, for the second element of `deletedIndexes` i.e 2, in the loop `for (var j = 0; j < rules.length; j++)`
No conditions are satisfied and the 2nd rule (i.e the rule currently having index 1) is not made invalid.
So in order to avoid this problem, a copy of the original _answerGroup was made and we compared the deleted indexes w.r.t to the copy
I hope the things are a bit more clear now.

Reply:
Ah, I see, thanks a lot for the explanation! That is very clear and I get it now. The issue is that we look through all the answer groups and rules for every deleted index but we need to make decisions based on the original value of inputs.x since the value of inputs.x gets modified over the course of the algorithm.

Could you please add a comment above the copy() statement to explain why the copy is necessary? I think that will be useful for future contributors.

Thanks again for the explanation!

Reply:
done